pkg/skaffold/build/gcb/status.go

```diff
@@ -0,0 +1,197 @@
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/api/cloudbuild/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/gcp"
+)
+
+var (
+	reporter     *statusReporterImpl
+	reporterOnce sync.Once
+	client       *cloudbuild.Service
+	clientOnce   sync.Once
+)
+
+type statusReporter interface {
+	getStatus(ctx context.Context, projectID string, buildID string) <-chan result
+}
+
+type statusReporterImpl struct {
+	results         map[jobID]chan result
+	resultMutex     sync.RWMutex
+	requests        chan jobRequest
+	pollingInterval time.Duration
+}
+
+type jobID struct {
+	projectID string
+	buildID   string
+}
+
+type jobRequest struct {
+	jobID
+	ctx context.Context
+}
+
+type result struct {
+	jobID
+	status *cloudbuild.Build
+	err    error
+}
+
+func getStatusReporter() statusReporter {
+	reporterOnce.Do(func() {
+		reporter = &statusReporterImpl{
+			pollingInterval: PollingInterval,
+			results:         make(map[jobID]chan result),
+			requests:        make(chan jobRequest, 100),
+		}
+
+		go reporter.start()
+	})
+	return reporter
+}
+
+func (r *statusReporterImpl) start() {
+	poll := time.NewTicker(r.pollingInterval)
+	jobsByProjectID := make(map[string]map[jobID]jobRequest)
+	for {
+		select {
+		case req := <-r.requests:
+			if _, found := jobsByProjectID[req.projectID]; !found {
+				jobsByProjectID[req.projectID] = make(map[jobID]jobRequest)
+			}
+			jobsByProjectID[req.projectID][req.jobID] = req
+		case <-poll.C:
+			for projectID, jobs := range jobsByProjectID {
+				for id, req := range jobs {
+					if req.ctx.Err() != nil {
+						r.setResult(id, result{jobID: id, err: req.ctx.Err()})
+						delete(jobs, id)
+					}
+				}
+				statuses, err := getStatuses(projectID, jobs)
+				if err != nil {
+					for id := range jobs {
+						r.setResult(id, result{jobID: id, err: err})
+						delete(jobs, id)
+					}
+					continue
+				}
+				for id := range jobs {
+					cb := statuses[id]
+					switch cb.Status {
+					case StatusQueued, StatusWorking, StatusUnknown:
+					case StatusSuccess:
+						r.setResult(id, result{jobID: id, status: cb})
+						delete(jobs, id)
+					case StatusFailure, StatusInternalError, StatusTimeout, StatusCancelled:
+						r.setResult(id, result{jobID: id, err: fmt.Errorf("cloud build failed: %s", cb.Status)})
+						delete(jobs, id)
+					default:
+						r.setResult(id, result{jobID: id, err: fmt.Errorf("unknown status: %s", cb.Status)})
+						delete(jobs, id)
+					}
+				}
+			}
+		}
+	}
+}
+
+func getStatuses(projectID string, jobs map[jobID]jobRequest) (map[jobID]*cloudbuild.Build, error) {
+	var err error
+	clientOnce.Do(func() {
+		client, err = cloudbuild.NewService(context.Background(), gcp.ClientOptions()...)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting cloudbuild client: %w", err)
+	}
+	backoff := NewStatusBackoff()
+	var cb *cloudbuild.ListBuildsResponse
+	if waitErr := wait.Poll(backoff.Duration, RetryTimeout, func() (bool, error) {
```

Comment:
I think the backoff should be handled by the caller to `getStatuses()`.  This will block any new requests that come in.

Reply:
changed. PTAL